pulpcore/plugin/repo_version_utils.py

```diff
@@ -31,21 +30,32 @@ def remove_duplicates(repository_version):
         repository_version: The :class:`~pulpcore.plugin.models.RepositoryVersion` to be checked
             and possibly modified.
     """
-    query_for_repo_duplicates_by_type = defaultdict(lambda: Q())
-    for item in repository_version.added():
-        detail_item = item.cast()
-        if detail_item.repo_key_fields == ():
+    added_content = repository_version.added()
+    existing_content = repository_version.content.exclude(pk__in=added_content)
+    repository = repository_version.repository.cast()
+    content_types = {ctype.get_pulp_type(): ctype for ctype in repository.CONTENT_TYPES}
+
+    for pulp_type in content_types.keys():
+        type_obj = content_types[pulp_type]
+        new_content_qs = type_obj.objects.filter(
+            pk__in=added_content.filter(pulp_type=pulp_type)
+        )
+
+        if type_obj.repo_key_fields == ():
             continue
-        unit_q_dict = {
-            field: getattr(detail_item, field) for field in detail_item.repo_key_fields
-        }
-        item_query = Q(**unit_q_dict) & ~Q(pk=detail_item.pk)
-        query_for_repo_duplicates_by_type[detail_item._meta.model] |= item_query
-
-    for model in query_for_repo_duplicates_by_type:
-        _logger.debug(_("Removing duplicates for type: {}".format(model)))
-        qs = model.objects.filter(query_for_repo_duplicates_by_type[model])
-        repository_version.remove_content(qs)
+
+        if new_content_qs.count():
+            find_dup_qs = Q()
+            for content_unit in new_content_qs:
```

Comment:
if you do:
```suggestion
            repo_key_fields_with_pk = type_obj.repo_key_fields + ('pk',)
            for content_unit in new_content_qs.values(*repo_key_fields_with_pk):
```
it would already bring the result in dicts, so you won't need to do a dict comprehension for each item
```python
pk = str(content_unit.pop("pk"))
item_query = Q(**content_unit) & ~Q(pk=pk)
```